### ISSUE TITLE
Enhance directory watch for shadow copy

### ISSUE THREAD

**Issue by MihaMarkic** [dotnet/aspnetcore#46629]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the solution you'd like

From what I understand, EnableShadowCopy watches for .dll changes in root directory only. I guess the main scenario is to allow deploying entire app each time.
In my case, I have a sub directory where I place some assemblies that can be/are replaced often. However, shadow copy doesn't trigger in my case. I'd like to add my sub directory to the list of changes that trigger shadow copy.
Another issue is that shadow copy is triggered only on .dll changes (assuming from experience). Adding an optional extension(s) might be useful

I guess something like (if more that one settings with same name is possible
```xml
<handlerSetting name="additionalDirectoryToWatch" value="./subdirectory/*.dll" />
<handlerSetting name="additionalDirectoryToWatch" value="./anotherdir/*.txt" />
```
or 
```xml
<handlerSetting name="additionalDirectoriesToWatch" value="./subdirectory/*dll;../anotherdir/*.txt" />
```

**Comment by BrennanConroy** [comment]
You can add a randomly named dll to trigger this behavior today. Or touch a current dll to update its timestamp.

It's unclear why you want random txt files to trigger an app restart.

We'll backlog for now and see if anymore asks for this type of feature.

**Comment by MihaMarkic** [comment]
@BrennanConroy Yep, I've figured out the two workarounds, I still think my solution would be more straightforward and cleaner. But it's not a big deal, since workarounds work.
As per .txt file - from what I understand, IIS shadow copies everything to shadow directory and all files are served/accessed from there. So, how does one update a non .dll file then? Or are non .dll files still served from original directory?

**Comment by BrennanConroy** [comment]
> Or are non .dll files still served from original directory?

I believe the webroot is still the original (non-shadow copy) folder. So any .txt changes will be seen by the app.

**Comment by MihaMarkic** [comment]
@BrennanConroy Are you sure? I'm asking because from what I'm seeing is that entire app directory is being shadow copied, not just .dlls. And hence I assume app will server local copies, not root ones.
IOW if I had to guess how it works: iis copies all app files to shadow directory and run app from there including all non .dlls. App serves content based on relative paths, so, from the app POV app is ignorant that it runs in a shadow dir.

**Comment by BrennanConroy** [comment]
The webroot is set to the original directory. You can easily test this by modifying some file (e.g. appsettings.json) and seeing that the copy in the shadow copy directory does nothing, but the copy in the original directory affects the app.

**Comment by MihaMarkic** [comment]
@BrennanConroy I've finally got some time and yes, you are correct that it loads (at least .txt but I guess all non .dll related) files from original directory. Curiously though, it'd still shadow copy them. Anyway, there is not need to trigger shadow copy on non .dll changes since it takes them from original directory.

**Comment by BrennanConroy** [comment]
> Anyway, there is not need to trigger shadow copy on non .dll changes since it takes them from original directory.

Am I understanding correctly, this means everything is working as you'd like it to and there isn't a need for `additionalDirectoryToWatch`?

**Comment by appelgran** [comment]
This touches another issue which @MihaMarkic already mentioned here. That shadow copying is done on the whole original directory. Currently this feature only makes sense for very small sites. Not for en Umbraco site with 11k files for example, [see this issue](https://github.com/dotnet/AspNetCore.Docs/issues/23733#issuecomment-1379675259). In .NET Framework 4 shadow copying was only done for assemblies and there was also an option to limit to specific folder(s) ([source, see ShadowCopyDirectories](https://learn.microsoft.com/en-us/dotnet/framework/app-domains/shadow-copy-assemblies)). It would make sense for it to work in the same way here and now.

And of course a changed .dll in a sub directory should trigger the same as a change in the root directory. That's a bug!

**Comment by luke-barnett** [comment]
I believe I have another use case that is caused by missing/extendable triggers for the copy.

I'm deploying a relatively stock standard Umbraco deployment to an infrastructure service (non-Azure), and I'm running into issues when I delete the relevant TEMP directory from the main wwwroot directory. Upon restarting the app, as there are no dll changes, it then seemingly fails at boot as it tries to attach a file watcher from the copy in the ShaowCopy folder to (presumably) the now missing folder in wwwroot.

This could be solved a couple of ways, I would probably prefer that it not to copy the entire directory or to add exclusions, but it could also concievably be solved by another trigger for the copy? But that would get incredibly messy if each time the app writes to disk for a search cache it restarts the app.

It could also very well just be a bug in it's own right that it should more gracefully handle a directory no longer present that was shadow copied earlier?

@BrennanConroy happy to take a lead here as to if this should just be a different issue to this current topic.

**Comment by abraluve** [comment]
I would like to see aswell subfolder to watch. E.g. I have plugin folder that is runtime loaded. Any changes that would trigger shadow copy would be nice.
